### ISSUE TITLE
Issue with the OLCI Level 1B RGB Daily Accumulated

### ISSUE THREAD

**Issue by randomorder** [geosolutions-it/eumetsat-EOWS#482]
Reported by CGI. JIRA ticket [here](https://cgi-bds.atlassian.net/browse/AH-11)

> Issue with the OLCI Level 1B RGB Daily Accumulated - Sentinel-3: the generation of the full resolution download takes 20-30 minutes to be ready.

We want to investigate and eventually provide an estimate to CGI

**Comment by randomorder** [comment]
I'll try to reproduce the issue now @simboss

**Comment by randomorder** [comment]
**NOTE**: you must provide a valid **access_token** in the request
```
curl 'https://oview.eumetsat.int/geoserver/ows?access_token=xxx' \
  -H 'Content-Type: text/plain' \
  --data-raw $'<?xml version="1.0" encoding="UTF-8"?>\n    <wps:Execute version="1.0.0" service="WPS"\n                xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"\n                xmlns="http://www.opengis.net/wps/1.0.0"\n                xmlns:wfs="http://www.opengis.net/wfs"\n                xmlns:wps="http://www.opengis.net/wps/1.0.0"\n                xmlns:ows="http://www.opengis.net/ows/1.1"\n                xmlns:gml="http://www.opengis.net/gml"\n                xmlns:ogc="http://www.opengis.net/ogc"\n                xmlns:wcs="http://www.opengis.net/wcs/1.1.1"\n                xmlns:xlink="http://www.w3.org/1999/xlink"\n                xsi:schemaLocation="http://www.opengis.net/wps/1.0.0 http://schemas.opengis.net/wps/1.0.0/wpsAll.xsd"\n    >\n        <ows:Identifier>gs:Download</ows:Identifier>\n        <wps:DataInputs>\n            <wps:Input>\n                <ows:Identifier>layerName</ows:Identifier>\n                <wps:Data>\n                    <wps:LiteralData>copernicus:daily_sentinel3ab_olci_l1_rgb_fulres</wps:LiteralData>\n                </wps:Data>\n            </wps:Input>\n            <wps:Input>\n                <ows:Identifier>filter</ows:Identifier>\n                <wps:Data>\n                    <wps:ComplexData mimeType="text/plain; subtype=cql">time = 2021-08-04T09:53:00.000Z</wps:ComplexData>\n                </wps:Data>\n            </wps:Input>\n            <wps:Input>\n                <ows:Identifier>outputFormat</ows:Identifier>\n                <wps:Data>\n                    <wps:LiteralData>image/tiff</wps:LiteralData>\n                </wps:Data>\n            </wps:Input>\n            \n    <wps:Input>\n    <ows:Identifier>writeParameters</ows:Identifier>\n    <wps:Data>\n        <wps:ComplexData xmlns:dwn="http://geoserver.org/wps/download">\n            <dwn:Parameters>\n                <dwn:Parameter key="compression">Deflate</dwn:Parameter>\n            </dwn:Parameters>\n        </wps:ComplexData>\n    </wps:Data>\n    </wps:Input>\n\n        </wps:DataInputs>\n        <wps:ResponseForm>\n            <wps:ResponseDocument storeExecuteResponse="true" status="true">\n                <wps:Output asReference="true" mimeType="application/zip">\n                    <ows:Identifier>result</ows:Identifier>\n                </wps:Output>\n            </wps:ResponseDocument>\n        </wps:ResponseForm>\n    </wps:Execute>' 
```


Or you can use the client directly. Link and credentials [here](https://docs.google.com/document/d/1phYbT3e1KQcwzcc1p7DCHBSg_imp21b4Nuzr12FF3Zw/edit#bookmark=kix.8dk0k4bqblo2)

**Comment by simboss** [comment]
@dromagnoli please, investigate first and then propose a fix and an estimate.

**Comment by dromagnoli** [comment]
Starting to investigate on that although I suppose the outcome will be very similar to what already found with the past investigation on WCS.
https://github.com/geosolutions-it/internal/issues/117

**Comment by simboss** [comment]
Let's see.

Once you are done, report here and we will see what to do.

**Comment by dromagnoli** [comment]
Some more outcomes on this, based on a previous investigation made on April for another dataset (RGB AirMass):
After that investigation we have improved the WPS-Download to support direct TIFF output without zipping compression and we have also added support for a couple of JAVA_OPTS to increase the internal buffer sizes for data writing and data copy (from WPS temporary computation to output result).

See details here:
https://docs.google.com/spreadsheets/d/11XpjZ8yU2q7lEgLOY2DkbDIOiWAomsnPxaLML67HoDY/edit#gid=0

Basically, switching the response output mimeType from using application/zip to image/tiff already provides a great improvement. (during that investigation, the throughput was improving of a 3x factor).

So replacing this:
` <wps:ResponseForm>
      <wps:ResponseDocument storeExecuteResponse="true" status="true">
         <wps:Output asReference="true" mimeType="application/zip">
            <ows:Identifier>result</ows:Identifier>
         </wps:Output>
      </wps:ResponseDocument>
   </wps:ResponseForm>`

with this:

` <wps:ResponseForm>
      <wps:ResponseDocument storeExecuteResponse="true" status="true">
         <wps:Output asReference="true" mimeType="image/tiff">
            <ows:Identifier>result</ows:Identifier>
         </wps:Output>
      </wps:ResponseDocument>
   </wps:ResponseForm>`

Note the mimeType being changed to image/tiff. This allows to skip the final zipping since the TIFF is already Deflate compressed.

We did a quick test with Alessandro on the Eumetsat server and the output has been produced in 4 minutes whilst the zip request was still running after 13 minutes (and thrown a PSQL error before finishing)

Note that the final output is around 2.4 GB so even increasing the size of the buffers for the writing and the copy, as instructed [here](https://docs.geoserver.org/stable/en/user/extensions/wps-download/rawDownload.html#writing-buffering-options) might slightly improve too (it would be great to test these flags effect on their infrastructure).

Finally, just out of curiosity of mine, I did a test with ZSTD compression which we have recently supported on main and stable.
On my machine, the same request get served in around 4 minutes with ZSTD instead of 8 minutes needed for DEFLATE.

Depending on the final usage of the downloaded data we might suggest to also offer ZSTD compression from the client and let have the data ZSTD compressed. 
This will require that the end user of the data has proper tools/libraries needed to support ZSTD decompression.

**Comment by randomorder** [comment]
Thanks @dromagnoli 

Do you happen to know what's the level of support for ZSTD by popular client application? Maybe they could leave Deflate as default and add a checkbox or a dropdown to allow for ZSTD
![image](https://user-images.githubusercontent.com/5264230/132470279-d14fcbd7-382e-4cce-ab4a-4c0c1f76b804.png)

The low hanging fruit is the change to the WPS requests to avoid ZIP compression for rasters. We can submit a small proposal to EUM to change this in the client application. The **drawback is of course the loss of the accompanying SLD file for the style**, we should state it clearly.

Concerning NetCDFs stores, they curently already being downloaded in a zipfile as GeoTiifs, not NetCDFs so changing the mimetype should not introduce any issue

We could benefit from a tuning of the buffer size but is going to take some time to find a good value and I am not 100% sure we can perform the activity in DEV where we have a different kind of infrastructure. Maybe we can ask for a small window in VAL to perform the task

Side note: To my knowledge they are still waiting for an estimate @simboss

**Comment by dromagnoli** [comment]
@randomorder 
about the accompanying SLD I'm wondering if the client could provide some more options like "skip zipping result" or something similar (to force raw geotiff request instead of zipping).
In that case, maybe the client could query GeoServer to get the default style name for the layer and if it's not "raster" (the default style for raster layer) report a warning like "Layer XYZ uses style ABC, skip zipping won't include it in the output" (sorry for the ugly phrasing :D)

About ZSTD I guess that, for example, recent QGIS versions already suppor it. Let me know if I need to investigate a bit more on which version are supporting it.

**Comment by simboss** [comment]
@dromagnoli we need one last extra effort to investiga on the possibility to perform a directo download that would skip any processing.
We can even think about a specific process if that helps.

Be creative.

**Comment by dromagnoli** [comment]
@simboss Yep. yesterday I started some investigation but I got buried in other tasks.
I should be able to get back on that investigation today or tomorrow.

**Comment by dromagnoli** [comment]
proposed plan:
https://docs.google.com/document/d/1SHQC8z8xKKSuFdwDZfFIqmyjWLJmdkGYmgq5fjaJ_XQ/edit#

will add some task-split and estimates
